### PR TITLE
do base->enhanced fuzzy wishlist matching at time of matching

### PR DESCRIPTION
**Pull request by nev-r**
instead of quadrupling the overall number of wishlist rolls,
this does fuzzy matching
- when deciding if a weapon matches recommended roll to a weapon
- when generating the InventoryWishListRoll
- when deciding to put a thumbs up on a perk circle

additionally, fixes an issue where `getWishListPlugs` returned a truncated set,
because `isWeaponOrArmorOrGhostMod` did not recognize enhanced perks (they're missing most useful ICHs)

so now enhanced perks actually get a thumbs up, on wishlist-matching rolls

### CODE REVIEW

File: src/app/item-popup/Plug.tsx [hunk 1]
```diff
@@ -130,6 +131,12 @@ export function PerkCircleWithTooltip({
       craftingData={socketInfo.craftingData?.[plug.plugDef.hash]}
     />
   );
+
+  const isRecommendedPerk =
+    wishlistRoll &&
+    (wishlistRoll.wishListPerks.has(plug.plugDef.hash) ||
+      wishlistRoll.wishListPerks.has(enhancedToPerk[plug.plugDef.hash]));
```
Comment (robojumper) [inline src/app/item-popup/Plug.tsx:138]: This is just `isWishListPlug` below?

File: src/app/wishlists/wishlists.ts [hunk 2]
```diff
@@ -96,18 +106,19 @@ function getWishListPlugs(item: DimItem, wishListRoll: WishListRoll): Set<number
...
+      // which has already ensured item.sockets exists
+      outer: for (const s of item.sockets!.allSockets) {
         if (s.plugOptions) {
-          for (const dp of s.plugOptions) {
-            if (dp.plugDef.hash === rp) {
+          for (const plug of s.plugOptions) {
+            if (
+              plug.plugDef.hash === recommendedPerk ||
+              perkToEnhanced[recommendedPerk] === plug.plugDef.hash
+            ) {
```
Comment (robojumper) [inline src/app/wishlists/wishlists.ts:121]: I know this is pre-existing but these nested labelled loops could probably be more concise if replaced with `every`/`some`.

**Review by robojumper — COMMENTED** [review]
Would be nice if this still added the enhanced perks to the armory view to indicate that DIM will actually match these.

File: src/app/wishlists/wishlists.ts (hunk 2 above)
Comment (nev-r) [reply src/app/wishlists/wishlists.ts:121]: i assumed this was done for performance or smth, i just kept the overarching structure and renamed it a bit. i think it's effectively a double, nested .some? i'd prefer to leave that to someone else to mess with.

File: src/app/item-popup/Plug.tsx (hunk 1 above)
Comment (nev-r) [reply src/app/item-popup/Plug.tsx:138]: slightly different typings but i could make the double .has a shared function yeah

Comment (nev-r) [reply src/app/item-popup/Plug.tsx]: should be good

**Comment by 48klocs** [comment]
Ah, this makes fine sense. Thanks!

**Pull request merged**